### ISSUE TITLE
Linux 5.15 exposes a single core instead of 4 on Hive Unmatched (riscv64 board)

### ISSUE THREAD

**Issue by zorun** [open-mpi/hwloc#536]
### What version of hwloc are you using?

Latest git (057fc67b3dd91fd1f251c39c3aa1465f18e83554)
```
./utils/lstopo/lstopo --version
lstopo 2.9.0a1-git
```

### Which operating system and hardware are you running on?

* Ubuntu 22.04 on riscv64
* `Linux gcc92 5.15.0-1014-generic #16-Ubuntu SMP Thu Jun 2 13:45:22 UTC 2022 riscv64 riscv64 riscv64 GNU/Linux`
* This is gcc92 here: https://cfarm.tetaneutral.net/machines/list/

```
$ ./utils/lstopo/lstopo -
Machine (16GB total)
  Package L#0
    NUMANode L#0 (P#0 16GB)
    L2 L#0 (2048KB) + Core L#0
      L1d L#0 (32KB) + L1i L#0 (32KB) + PU L#0 (P#0)
      L1d L#1 (32KB) + L1i L#1 (32KB) + PU L#1 (P#1)
      L1d L#2 (32KB) + L1i L#2 (32KB) + PU L#2 (P#2)
      L1d L#3 (32KB) + L1i L#3 (32KB) + PU L#3 (P#3)
  HostBridge
    PCIBridge
      PCIBridge
        PCIBridge
          PCI 06:00.0 (NVMExp)
            Block(Disk) "nvme0n1"
  Block "mmcblk0"
  Net "eth0"
```

### Details of the problem

* hwloc detects a single core with 4 threads:

```
$ ./utils/hwloc/hwloc-calc -N core all
1
$ ./utils/hwloc/hwloc-calc -N pu all
4
```

* as far as I understand the hardware, it's supposed to have 4 cores

### Additional information

You normally already have access to the gcc farm for remote debugging.

**Comment by bgoglin** [comment]
Hello. I wondered the same when testing the gcc farm yesterday. Linux is showing a single core with 4 threads. I saw some possibly related patched on the Linux kernel mailing list today, I'll ask if they could fix this kernel bug.

**Comment by zorun** [comment]
Right, thanks!

Note that gcc91 has the correct amount of cores detected (2), but it's a different SoC and it's running with a newer kernel (5.18.5)

**Comment by bgoglin** [comment]
Reported the issue to RISC-V developers at https://lkml.org/lkml/2022/7/5/1568 and there are now kernel patches at https://lore.kernel.org/linux-riscv/20220705190435.1790466-1-mail@conchuod.ie/ which should fix the issue.

**Comment by ConchuOD** [comment]
Oh dear, I didn't realise that I would spam /your/ issue by doing that. Good ole overzealous github...

**Comment by bgoglin** [comment]
Don't worry @ConchuOD , that's fine, we can unsubscribe from the issue if needed (but I won't).

**Comment by zorun** [comment]
Thanks a lot for the detective work and the quick patch!

And sorry about the misleading comment on the VisionFive, I totally forgot that we're using a downstream kernel...

**Comment by ConchuOD** [comment]
> Thanks a lot for the detective work and the quick patch!

My hope is to get the addition of defaults backported & the cpu-map stuff into v5.20 (but not backport it).

> And sorry about the misleading comment on the VisionFive, I totally forgot that we're using a downstream kernel...

Ye I knew there was a vendor kernel but I completely forgot about it too.. dw about it :)

**Comment by ConchuOD** [comment]
Just as an update, this has missed v6.0 due as it didn't get reviews from the RISC-V side in time. I will resubmit based on v6.0-rc1.

**Comment by ConchuOD** [comment]
The actual fix for this is in mainline (as of like 5 minutes ago), when is this considered "fixed"? Once the backports have been done?

**Comment by bgoglin** [comment]
I am fine with marking as fixed now or later, as you want. I'll keep a note about which kernels contain the fix in case the issue is reported again before the fix reaches downstream distributions.

**Comment by ConchuOD** [comment]
In that case, I'll mark it fixed when I see the backports go through. Thanls.

**Comment by bgoglin** [comment]
Looks like the fix has reached 6.1 and been backported to stables series, I am closing this issue, thanks.

**Comment by ConchuOD** [comment]
> Looks like the fix has reached 6.1 and been backported to stables series, I am closing this issue, thanks.

Yeah, although I don't think the stable kernels have technically been released I think closing it is fair at this point. Thanks!
